src/brevitas/graph/equalize.py

```diff
@@ -350,57 +350,76 @@ def _is_reshaping_op(node: Node) -> bool:
+        elif _is_scale_invariant_module(
+                graph_model, node) or _is_scale_invariant_function(node) or _is_reshaping_op(node):
+            state = find_srcs(graph_model, node, state)
+            state = find_sinks(graph_model, node, state)
         elif (node.op == 'call_method' and node.target in _residual_methods or
               node.op == 'call_function' and node.target in _residual_fns):
-            walk_region(graph_model, node, history, srcs, sinks, walk_forward=True)
-            walk_region(graph_model, node, history, srcs, sinks, walk_forward=False)
-        elif _is_reshaping_op(node):
-            walk_region(graph_model, node, history, srcs, sinks, walk_forward=walk_forward)
+            state = find_srcs(graph_model, node, state)
+            state = find_sinks(graph_model, node, state)
+        else:
+            # If we meet an unrecognized op, we add None to invalidate the region
+            state['srcs'].add(None)
+    return state
+
+
+def find_sinks(graph_model: GraphModule, starting_node: Node, state) -> Dict[str, Set]:
+    node_list = starting_node.users
+    for node in node_list:
+        # we keep a history of how the graph has been walked already, invariant to the direction,
+        # to avoid getting stuck in a loop
+        # Note that the path is inverted with respect to find_srcs
+        path = (starting_node, node)
+        if path not in state['history']:
+            state['history'].add(path)
         else:
             continue
+        if _is_supported_module(graph_model, node):
+            module = get_module(graph_model, node.target)
+            # It is not possible to equalize through LayerNorm as sink
+            if not isinstance(module, nn.LayerNorm):
+                state['sinks'].add(node.target)
+        elif _is_scale_invariant_module(
+                graph_model, node) or _is_scale_invariant_function(node) or _is_reshaping_op(node):
+            state = find_sinks(graph_model, node, state)
+        elif (node.op == 'call_method' and node.target in _residual_methods or
+              node.op == 'call_function' and node.target in _residual_fns):
+            state = find_sinks(graph_model, node, state)
+            state = find_srcs(graph_model, node, state)
+        else:
+            # If we meet an unrecognized op, we add None to invalidate the region
+            state['sinks'].add(None)
+    return state
 
 
 def _extract_regions(graph_model: GraphModule) -> Set[Tuple[str]]:
     regions = set()
     for node in graph_model.graph.nodes:
         if _is_supported_module(graph_model, node):
-            srcs, sinks = {node.target}, set()
-            walk_region(graph_model, node, set(), srcs, sinks, walk_forward=True)
-            if sinks:
+            state = dict()
+            state['srcs'] = {node.target}
+            state['sinks'] = set()
+            state['history'] = set()
+            state = find_sinks(graph_model, node, state)
```

Comment:
do we need to reassign the state variable?

Reply:
No I did it mostly for clarity, to make clear that it gets changed and updated. I would rather keep it if it's not too ugly or non-pythonic